lms/services/grouping.py

```diff
@@ -57,27 +57,32 @@ def upsert_with_parent(
             # Make sure we have a PK for the parent before upserting
             self._db.flush()
 
+        values = [
+            {
+                # Things we generate
+                "application_instance_id": self.application_instance.id,
+                "authority_provided_id": self.generate_authority_provided_id(
+                    self.application_instance.tool_consumer_instance_guid,
+                    grouping["lms_id"],
+                    parent,
+                    type_,
+                ),
+                "updated": func.now(),
+                # From params
+                "parent_id": parent.id,
+                "type": type_,
+                # Things the caller provides
+                "lms_id": grouping["lms_id"],
+                "lms_name": grouping["lms_name"],
+                "extra": grouping.get("extra"),
+            }
+            for grouping in grouping_dicts
+        ]
```

Comment:
Moving this out makes debugging a lot easier, as you can drop a break point after all this is done, instead of getting lead into all the stuff that happens here.